### PR TITLE
region error messages involving impls are confusing 

### LINKED ISSUE #52742: region error messages involving impls are confusing
```rust
#![feature(nll)]
#![feature(in_band_lifetimes)]
#![feature(impl_header_lifetime_elision)]

struct Foo<'a, 'b> {
    x: &'a u32,
    y: &'b u32,
}

struct Bar<'b> {
    z: &'b u32
}

impl Foo<'_, '_> {
    fn take_bar(&mut self, b: Bar<'_>) {
        self.y = b.z
    }
}

fn main() { }
``` 

produces this confusing error:

```
error[E0621]: explicit lifetime required in the type of `b`
  --> src/main.rs:15:9
   |
14 |     fn take_bar(&mut self, b: Bar<'_>) {
   |                            - consider changing the type of `b` to `Bar<'_>`
15 |         self.y = b.z
   |         ^^^^^^^^^^^^ lifetime `'_` required
```

I think the problem here is that the "nice region error" code is actually getting in the way of the NLL errors, which are better.

### PR DESCRIPTION
**Pull request by davidtwco**
Part of #52742.

r? @nikomatsakis

### CODE REVIEW

**Comment by kennytm** [comment]
@bors r-

```diff
 failures:
 ---- [ui] ui\nll\issue-52742.rs stdout ----
 diff of stderr:
-	error: unsatisfied lifetime constraints
-	  --> $DIR/issue-52742.rs:25:9
+	error[E0106]: missing lifetime specifiers
+	  --> $DIR/issue-52742.rs:23:10
3	   |
-	LL |     fn take_bar(&mut self, b: Bar<'_>) {
-	   |                 ---------         -- let's call this `'1`
-	   |                 |
-	   |                 has type `&mut Foo<'_, '2>`
-	LL |         self.y = b.z
-	   |         ^^^^^^^^^^^^ requires that `'1` must outlive `'2`
+	LL | impl Foo<'_, '_> {
+	   |          ^^ expected 2 lifetime parameters
10	
11	error: aborting due to previous error
12	
+	For more information about this error, try `rustc --explain E0106`.
13	
```

File: src/test/ui/nll/issue-52742.rs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(nll)]
+#![feature(in_band_lifetimes)]
```
Comment (nikomatsakis) [inline src/test/ui/nll/issue-52742.rs]: the problem here is that we now need another feature gate here

```
#![feature(impl_header_lifetime_elision)]
```
```

Comment (davidtwco) [reply src/test/ui/nll/issue-52742.rs]: Fixed.

**Pull request merged**